object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -338,44 +358,144 @@ private void handleAssignment(AssignmentNode node, Set<LocalVarWithTree> newDefs
+          if (n != null && isVarInDefs(newDefs, n)) {
+            ImmutableSet<LocalVarWithTree> immutableSet =
+                getSetContainingAssignmentTreeOfVar(newDefs, n);
+            ImmutableSet<LocalVarWithTree> newImmutableSet =
+                FluentIterable.from(immutableSet)
+                    .append(
+                        new LocalVarWithTree(
+                            new LocalVariable((LocalVariableNode) lhs), node.getTree()))
+                    .toSet();
+            newDefs.remove(immutableSet);
+            newDefs.add(newImmutableSet);
+          }
+        } else {
+          LocalVarWithTree lhsLocalVarWithTreeNew =
+              new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+          newDefs.add(ImmutableSet.of(lhsLocalVarWithTreeNew));
+        }
       }
 
       // Ownership Transfer
       if (rhs instanceof LocalVariableNode && isVarInDefs(newDefs, (LocalVariableNode) rhs)) {
         // If the rhs is a LocalVariableNode that exists in the newDefs (Note that if a
         // localVariableNode exists in the newDefs it means it isn't assigned to a null
         // literals), then it adds the localVariableNode to the newDefs
-        newDefs.add(
-            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree()));
-        newDefs.remove(getAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs));
+        ImmutableSet<LocalVarWithTree> setContainingRhs =
+            getSetContainingAssignmentTreeOfVar(newDefs, (LocalVariableNode) rhs);
+        LocalVarWithTree lhsLocalVarWithTreeNew =
+            new LocalVarWithTree(new LocalVariable((LocalVariableNode) lhs), node.getTree());
+        newDefs.add(ImmutableSet.of(lhsLocalVarWithTreeNew));
+        newDefs.remove(setContainingRhs);
+      }
+    }
+    return newDefs;
+  }
+
+  /**
+   * This method tries to find a local variable passed as a @MustCallChoice parameter. In the base
+   * case, if {@code node} is a local variable, it just gets returned. Otherwise, if node is a call
+   * (or a call wrapped in a cast), the code finds the parameter passed in the @MustCallChoice
+   * position, and recurses on that parameter.
+   *
+   * @param node
+   * @return {@code node} iff {@code node} represents a local variable, otherwise null
+   */
+  private @Nullable LocalVariableNode getLocalPassedAsMustCallChoiceParam(Node node) {
+    if (node instanceof TypeCastNode) {
+      node = ((TypeCastNode) node).getOperand();
+    }
+
+    if (node instanceof MethodInvocationNode || node instanceof ObjectCreationNode) {
+      List<Node> arguments = getArgumentsOfMethodOrConstructor(node);
+      List<? extends VariableElement> formals = getFormalsOfMethodOrConstructor(node);
+
+      for (int i = 0; i < arguments.size(); i++) {
+        if (typeFactory
+                .getTypeFactoryOfSubchecker(MustCallChecker.class)
+                .getDeclAnnotationNoAliases(formals.get(i), MustCallChoice.class)
```

Comment:
I think this should be a call to `ObjectConstructionAnnotatedTypeFactory#hasMustCallChoice` rather than a direct check via the must-call checker